File: src/main/java/com/github/morningzeng/toolset/dialog/AbstractPropDialog.java
package com.github.morningzeng.toolset.dialog;

import com.beust.jcommander.internal.Maps;
import com.fasterxml.jackson.core.type.TypeReference;
import com.github.morningzeng.toolset.Constants.IconC;
import com.github.morningzeng.toolset.action.SingleTextFieldDialogAction;
import com.github.morningzeng.toolset.component.ActionBar;
import com.github.morningzeng.toolset.component.Tree;
import com.github.morningzeng.toolset.dialog.AbstractPropDialog.AbstractRightPanel;
import com.github.morningzeng.toolset.dialog.AsymmetricPropDialog.RightPanel;
import com.github.morningzeng.toolset.model.Children;
import com.github.morningzeng.toolset.support.ScrollSupport;
import com.github.morningzeng.toolset.utils.ActionUtils;
import com.github.morningzeng.toolset.utils.ScratchFileUtils;
import com.intellij.icons.AllIcons.General;
import com.intellij.openapi.actionSystem.ActionUpdateThread;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.ui.JBSplitter;
import com.intellij.ui.components.JBLabel;
import com.intellij.ui.components.JBPanel;
import com.intellij.ui.components.JBPanelWithEmptyText;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.Action;
import javax.swing.BoxLayout;
import javax.swing.JComponent;
import javax.swing.SwingConstants;
import javax.swing.tree.DefaultMutableTreeNode;
import javax.swing.tree.TreeNode;
import java.awt.Dimension;
import java.awt.event.ActionEvent;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Consumer;
import java.util.function.Predicate;
import java.util.stream.Stream;

/**
 * @author Morning Zeng
 * @since 2024-05-27
 */
public abstract sealed class AbstractPropDialog<T extends Children<T>, P extends AbstractRightPanel<T>> extends DialogWrapper implements DialogSupport
        permits AsymmetricPropDialog, HashPropDialog, JWTPropDialog, SymmetricPropDialog {

    static final JBPanel<JBPanelWithEmptyText> EMPTY_PANEL = new JBPanelWithEmptyText();
    final Project project;
    final JBSplitter pane = new JBSplitter(false, "prop-dialog-splitter", .3f);
    final Tree<T> tree = new Tree<>();
    private final Consumer<List<T>> okAfterConsumer;
    private final Map<T, P> rightPanelMap = Maps.newHashMap();
    final AnAction addActions = ActionUtils.drawerActions("Add Item", "New create crypto prop item", General.Add, this.initGroupAction());
    final ActionBar actionBar = new ActionBar(this.barActions());

    @SuppressWarnings("unused")
    protected AbstractPropDialog(@Nullable final Project project) {
        this(project, symmetricCryptoProps -> {
        });
    }

    protected AbstractPropDialog(@Nullable final Project project, final Consumer<List<T>> okAfterConsumer) {
        super(project);
        this.project = project;
        this.okAfterConsumer = okAfterConsumer;

        this.tree.clearSelectionIfClickedOutside();
        this.tree.setNodes(ScratchFileUtils.read(this.typeReference()), Children::isGroup);
        this.tree.addTreeSelectionListener(e -> {
            final DefaultMutableTreeNode selectNode = (DefaultMutableTreeNode) this.tree.getLastSelectedPathComponent();
            if (Objects.nonNull(selectNode)) {
                final T value = tree.getNodeValue(selectNode);
                if (!this.enabledNode().test(value)) {
                    this.tree.clearSelection();
                    this.pane.setSecondComponent(EMPTY_PANEL);
                    return;
                }
            }
            this.defaultRightPanel();
        });
        this.tree.cellRenderer(prop -> {
            final JBLabel label = new JBLabel(prop.name(), prop.icon(), SwingConstants.LEFT);
            label.setEnabled(this.enabledNode().test(prop));
            return label;
        });
        this.actionBar.setLayout(new BoxLayout(this.actionBar, BoxLayout.LINE_AXIS));
    }

    AnAction[] barActions() {
        return new AnAction[]{this.addActions, this.deleteAction()};
    }

    AnAction deleteAction() {
        return new AnAction(IconC.REMOVE_RED) {
            @Override
            public void actionPerformed(@NotNull final AnActionEvent e) {
                delete();
            }
        };
    }

    Predicate<T> enabledNode() {
        return t -> true;
    }

    @Override
    public void delete() {
        this.tree.delete(treePaths -> this.defaultRightPanel());
    }

    @Override
    protected JComponent createCenterPanel() {
        this.createLeftPanel();
        this.defaultRightPanel();
        this.pane.setMinimumSize(new Dimension(700, 500));
        this.pane.setDividerWidth(1);
        return pane;
    }

    @Override
    protected void doOKAction() {
        super.doOKAction();
    }

    @Override
    protected Action @NotNull [] createActions() {
        return Stream.concat(
                Stream.of(new DialogWrapperAction("Apply") {
                    @Override
                    protected void doAction(final ActionEvent e) {
                        applyProp();
                    }
                }),
                Stream.of(super.createActions())
        ).toArray(Action[]::new);
    }

    private void applyProp() {
        this.rightPanelMap.forEach(this::writeProp);
        this.tree.reloadTree((TreeNode) this.tree.getLastSelectedPathComponent());
        ScratchFileUtils.write(this.tree.data(), this.typeReference());
        this.okAfterConsumer.accept(this.tree.data());
    }

    AnAction[] initGroupAction() {
        return new AnAction[]{
                new SingleTextFieldDialogAction(
                        "Group", "Add Group", name -> this.tree.create(this.generateBean(name, true), true)
                ) {
                    @Override
                    public @NotNull ActionUpdateThread getActionUpdateThread() {
                        return Optional.of(super.getActionUpdateThread())
                                .filter(ActionUpdateThread.BGT::equals)
                                .orElse(ActionUpdateThread.EDT);
                    }

                    @Override
                    public void update(@NotNull final AnActionEvent e) {
                        e.getPresentation().setEnabled(tree.isSelectionEmpty());
                    }
                },
                new AnAction("KeyPair") {
                    @Override
                    public void actionPerformed(@NotNull final AnActionEvent e) {
                        final T cryptoProp = generateBean("", false);
                        tree.create(cryptoProp, true);
                        defaultRightPanel();
                    }
                }
        };
    }

    abstract TypeReference<List<T>> typeReference();

    abstract T generateBean(final String name, final boolean isGroup);

    abstract void writeProp(final T prop, final P rightPanel);

    abstract P createRightPanel(final T t);

    void defaultRightPanel() {
        final T t = this.tree.getSelectedValue();
        if (Objects.isNull(t)) {
            this.pane.setSecondComponent(EMPTY_PANEL);
            return;
        }
        final P rightPanel = this.rightPanelMap.computeIfAbsent(t, this::createRightPanel);
        this.pane.setSecondComponent(rightPanel);
    }

    void createLeftPanel() {
        final JBPanel<JBPanelWithEmptyText> leftPanel = new JBPanel<>();
        leftPanel.setLayout(new BoxLayout(leftPanel, BoxLayout.Y_AXIS));

        leftPanel.add(this.actionBar);

        // add scroll
        leftPanel.add(ScrollSupport.getInstance(this.tree).verticalAsNeededScrollPane());

        this.pane.setFirstComponent(leftPanel);
    }

    protected sealed static abstract class AbstractRightPanel<T extends Children<T>> extends JBPanel<JBPanelWithEmptyText> permits RightPanel, HashPropDialog.RightPanel, JWTPropDialog.RightPanel, SymmetricPropDialog.RightPanel {
        protected T t;

        protected AbstractRightPanel(final T t) {
            this.t = t;
        }
    }
}

File: src/main/kotlin/com/github/morningzeng/toolbox/ui/dialog/AbstractPropDialog.kt
package com.github.morningzeng.toolbox.ui.dialog

import com.fasterxml.jackson.core.type.TypeReference
import com.github.morningzeng.toolbox.Constants
import com.github.morningzeng.toolbox.model.Children
import com.github.morningzeng.toolbox.ui.ScrollSupport
import com.github.morningzeng.toolbox.ui.action.SingleTextFieldDialogAction
import com.github.morningzeng.toolbox.ui.bar.ActionBar
import com.github.morningzeng.toolbox.ui.component.Tree
import com.github.morningzeng.toolbox.utils.ActionUtils
import com.github.morningzeng.toolbox.utils.ExpandMethodUtils.labelWidth
import com.github.morningzeng.toolbox.utils.GridBagUtils
import com.github.morningzeng.toolbox.utils.ScratchFileUtils
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.ui.LabeledComponent
import com.intellij.ui.JBSplitter
import com.intellij.ui.components.JBLabel
import com.intellij.ui.components.JBPanel
import com.intellij.ui.components.JBPanelWithEmptyText
import com.intellij.ui.components.JBTextField
import fleet.util.computeIfAbsentShim
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.event.ActionEvent
import java.util.function.Predicate
import javax.swing.Action
import javax.swing.BoxLayout
import javax.swing.JComponent
import javax.swing.SwingConstants
import javax.swing.tree.DefaultMutableTreeNode

/**
 * @author Morning Zeng
 * @since 2025-05-19
 */
abstract class AbstractPropDialog<T : Children<T>, P : AbstractPropDialog.AbstractRightPanel<T>>(
    protected val project: Project,
    protected val okAfter: (ts: MutableList<T>) -> Unit = {},
    protected val selectedAfter: (t: T) -> Unit = {}
) : DialogWrapper(project) {

    protected val tree: Tree<T> = Tree()

    private val emptyPanel = JBPanelWithEmptyText()
    private val pane = JBSplitter(false, "prop-dialog-splitter", .3f)
    private val rightPanelMap: MutableMap<T, P> = mutableMapOf()
    private val addActions: AnAction = ActionUtils.drawerActions(
        "Add Item", "New create crypto prop item", Constants.IconC.ADD_DRAWER, *initGroupAction()
    )
    private val actionBar = ActionBar(*barActions())

    init {
        tree.clearSelectionIfClickedOutside()
        val ts = ScratchFileUtils.read(typeReference())?.apply {
            sortWith(Children.comparable())
            forEach { it.children.sortWith(Children.comparable()) }
        } ?: mutableListOf()
        tree.setNodes(ts) { it.directory }
        tree.addTreeSelectionListener {
            tree.lastSelectedPathComponent?.let { it as DefaultMutableTreeNode }
                ?.also {
                    val value = tree.getNodeValue(it)
                    if (!enabledNode().test(value)) {
                        tree.clearSelection()
                        pane.secondComponent = emptyPanel
                        return@also
                    }
                }
            defaultRightPanel()
        }
        tree.cellRenderer {
            JBLabel(it.name(), it.icon(), SwingConstants.LEFT).apply {
                isEnabled = enabledNode().test(it)
            }
        }
        actionBar.layout = BoxLayout(actionBar, BoxLayout.LINE_AXIS)
    }

    open fun barActions(): Array<AnAction> {
        return arrayOf(addActions, deleteAction())
    }

    fun deleteAction(): AnAction {
        return object : AnAction(Constants.IconC.REMOVE_RED) {
            override fun actionPerformed(e: AnActionEvent) = delete()
        }
    }

    fun delete() {
        tree.delete {
            defaultRightPanel()
        }
    }

    open fun enabledNode(): Predicate<T?> = Predicate { true }

    override fun createCenterPanel(): JComponent {
        createLeftPanel()
        defaultRightPanel()
        pane.minimumSize = Dimension(700, 500)
        pane.dividerWidth = 3
        return pane
    }

    override fun doOKAction() {
        super.doOKAction()
        applyProp()
        tree.getSelectedValue()?.let { selectedAfter(it) }
    }

    override fun createActions(): Array<out Action?> {
        return arrayOf(
            object : DialogWrapperAction("Apply") {
                override fun doAction(e: ActionEvent?) {
                    applyProp()
                    tree.getSelectedValue()?.let { selectedAfter(it) }
                }
            },
            *super.createActions()
        )
    }

    private fun applyProp() {
        rightPanelMap.forEach { (t, p) -> writeProp(t, p) }
        tree.lastSelectedPathComponent?.let { it as DefaultMutableTreeNode }
            ?.also { tree.reloadTree(it) }
        ScratchFileUtils.write(tree.data(), typeReference())
        okAfter(tree.data())
        tree.getSelectedValue()?.let { selectedAfter(it) }
    }

    fun initGroupAction(): Array<AnAction> {
        return arrayOf(
            object : SingleTextFieldDialogAction(
                "Group", { name -> this.tree.create(this.generateBean(name, true), true) }, "Add Group"
            ) {
                override fun getActionUpdateThread(): ActionUpdateThread {
                    return super.getActionUpdateThread()
                }

                override fun update(e: AnActionEvent) {
                    e.presentation.isEnabled = tree.isSelectionEmpty
                }
            },
            object : AnAction("KeyPair") {
                override fun actionPerformed(e: AnActionEvent) {
                    val cryptoProp: T = generateBean("", false)
                    tree.create(cryptoProp, true)
                    defaultRightPanel()
                }
            }
        )
    }

    abstract fun typeReference(): TypeReference<MutableList<T>>

    abstract fun generateBean(name: String, isGroup: Boolean): T

    abstract fun writeProp(prop: T, rightPanel: P)

    abstract fun createRightItemPanel(t: T): P

    fun defaultRightPanel() {
        val t = tree.getSelectedValue()
        if (t == null) {
            pane.secondComponent = emptyPanel
            return
        }
        rightPanelMap.computeIfAbsentShim(t) { createRightItemPanel(t) }.also {
            pane.secondComponent = it
        }
    }

    fun createLeftPanel() {
        GridBagUtils.builder()
            .fill(GridBagUtils.GridBagFill.HORIZONTAL)
            .row { it.cell().weightX(1.0).add(actionBar) }
            .row {
                it.fill(GridBagUtils.GridBagFill.BOTH).cell().weightY(1.0).add(
                    ScrollSupport.getInstance(tree).verticalAsNeededScrollPane()
                )
            }
            .build()
            .apply { pane.firstComponent = this }
    }

    abstract class AbstractRightPanel<T : Children<T>>(
        protected val prop: T,
    ) : JBPanel<JBPanelWithEmptyText>() {

        private val emptyPanel = JBPanelWithEmptyText()
        protected val labelWidth = 50
        val titleTextField = LabeledComponent.create<JBTextField>(
            JBTextField(prop.name()), "Name", BorderLayout.WEST
        ).apply { labelWidth(labelWidth) }

        protected fun initLayout() {
            GridBagUtils.builder(this).apply {
                if (prop.directory) {
                    row {
                        it.fill(GridBagUtils.GridBagFill.HORIZONTAL)
                            .cell().weightX(1.0).add(titleTextField)
                    }.row {
                        it.fill(GridBagUtils.GridBagFill.BOTH)
                            .cell().weightX(1.0).weightY(1.0).add(emptyPanel)
                    }
                } else itemLayout()(this)
            }
        }

        protected abstract fun itemLayout(): (GridBagUtils.GridBagBuilder<AbstractRightPanel<T>>) -> Unit
    }

}